---
title: "Class 19:Vaccine Mini-Project"
author: "Suzanne Enos"
format: pdf
---
# Web scraping

Extracting the CDC figures for Pertussis cases in the US:
https://www.cdc.gov/pertussis/surv-reporting/cases-by-year.html
Using datapasta to paste data from website as a datafram.

```{r echo = FALSE}
cdc <- data.frame(
                                 Year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L),
         Cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617)
       )
```

```{r}
plot(cdc)
```

```{r}
library(ggplot2)

base <- ggplot(cdc) +
  aes(Year, Cases) +
  geom_point() +
  geom_line() +
  labs(title = "Pertussis Cases by Year (1922-2019)", y = "Number of Cases") +
  scale_y_continuous(labels = scales::label_comma())
base
```
First vaccine in 1946
```{r}
base + 
  geom_vline(xintercept = 1946, col = "blue", linetype = 2)
```
New vaccine formulation in 1996 - wP to aP (whole-cell to acellular)
US and many other countries switched. Not all countries switched.

```{r}
base + 
  geom_vline(xintercept = 1946, col = "blue", linetype = 2) +
  geom_vline(xintercept = 1996, col = "red", linetype = 2)
```
# Exploring CMI-PB data

Why is this vaccine-preventable disease on the upswing?

CMI-PB resource API returns JSON format. We will use **jsonlite**

```{r}
library(jsonlite)

subject <- read_json("http://cmi-pb.org/api/subject", simplifyVector = TRUE)
head(subject)
```

> How many wP and aP subjects are there?

```{r}
table(subject$infancy_vac)
```

> How many female non-white individuals are there in the dataset?

```{r}
table(subject$race, subject$biological_sex)
```

Let's look at the specimen table

```{r}
specimen <- read_json("http://cmi-pb.org/api/specimen", simplifyVector = T)
head(specimen)
dim(specimen)
```

```{r}
dim(subject)
```

```{r}
library(dplyr)
meta <- inner_join(specimen, subject)
dim(meta)
```

```{r}
titer <- read_json("http://cmi-pb.org/api/ab_titer", simplifyVector = T)
head(titer)
dim(titer)
```

> How many different isotypes are there?

```{r}
table(titer$isotype)
```

Merge titer and meta

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
```

```{r}
head(abdata)
```

> Q. What do you notice about "visit" number 8?

```{r}
table(abdata$visit)
```

Visit 8 still ongoing, still collecting data. Will just analyze visits 1-7.

# Examine IgG1 Ab titer levels

```{r}
ig1 <- filter(abdata, isotype == "IgG1", visit!= 8)
dim(ig1)
```
```{r}
head(ig1)
```

```{r}
table(abdata$antigen)
```

Analysis of the whole dataset: antigen levels (plot of antigen vs MFI)

```{r}
ggplot(ig1) +
  aes(MFI, antigen) +
  geom_boxplot()
```

Add faceting by visit

```{r}
ggplot(ig1) +
  aes(MFI, antigen) +
  geom_boxplot() +
  facet_wrap(vars(visit), nrow = 2)
```



```{r}
ggplot(ig1) +
  aes(MFI, antigen, col = ig1$infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit), nrow = 2)
```

```{r}
ggplot(ig1) +
  aes(MFI, antigen) +
  geom_boxplot() +
  facet_wrap(vars(infancy_vac, visit), nrow = 2)
```


Unclear what the difference in aP and wP

```{r}
filter(ig1, antigen== "FIM2/3") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  labs(title = "FIM2/3 antigen level") +
  theme_bw()
```

```{r}
filter(ig1, antigen== "PT") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  labs(title = "PT antigen level") +
  theme_bw()
```

> Do you see any clear differences in wP and aP response?

Not really


For RNA-Seq data the API query mechanism quickly hits the web browser interface limit for file size. We will present alternative download mechanisms for larger CMI-PB datasets in the next section. However, we can still do “targeted” RNA-Seq querys via the web accessible API.

For example we can obtain RNA-Seq results for a specific ENSEMBLE gene identifier or multiple identifiers combined with the & character:

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
head(rna)
```

Join rna to meta because different specimen from Ab titer

```{r}
ssrna <- inner_join(rna, meta)
head(ssrna)
```

```{r}
dim(ssrna)
```
Year of birth not age because age would be different for different visits.

Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).
```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2) + 
  theme_bw()
```

Expression peaks at visit 4, before Ab titer peak at visit 5 or 6.